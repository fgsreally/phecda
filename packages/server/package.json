{
  "name": "phecda-server",
  "version": "5.0.0-beta.20",
  "description": "provide express middlewares, `nestjs` format",
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "bin": {
    "phecda": "bin/cli.js"
  },
  "scripts": {
    "dev": "tsup --watch",
    "build": "tsup"
  },
  "exports": {
    ".": {
      "import": "./dist/index.mjs",
      "require": "./dist/index.js",
      "node": "./dist/index.js"
    },
    "./test": {
      "import": "./dist/test.mjs",
      "require": "./dist/test.js"
    },
    "./h3": {
      "import": "./dist/server/h3/index.mjs",
      "require": "./dist/server/h3/index.js"
    },
    "./express": {
      "import": "./dist/server/express/index.mjs",
      "require": "./dist/server/express/index.js"
    },
    "./koa": {
      "import": "./dist/server/koa/index.mjs",
      "require": "./dist/server/koa/index.js"
    },
    "./fastify": {
      "import": "./dist/server/fastify/index.mjs",
      "require": "./dist/server/fastify/index.js"
    },
    "./redis": {
      "import": "./dist/rpc/redis/index.mjs",
      "require": "./dist/rpc/redis/index.js"
    },
    "./kafka": {
      "import": "./dist/rpc/kafka/index.mjs",
      "require": "./dist/rpc/kafka/index.js"
    },
    "./rabbitmq": {
      "import": "./dist/rpc/rabbitmq/index.mjs",
      "require": "./dist/rpc/rabbitmq/index.js"
    },
    "./register": "./register/index.mjs",
    "./*": "./*"
  },
  "typesVersions": {
    "*": {
      ".": [
        "dist/index.d.ts"
      ],
      "test": [
        "dist/test.d.ts"
      ],
      "h3": [
        "dist/server/h3/index.d.ts"
      ],
      "express": [
        "dist/server/express/index.d.ts"
      ],
      "koa": [
        "dist/server/koa/index.d.ts"
      ],
      "fastify": [
        "dist/server/fastify/index.d.ts"
      ],
      "redis": [
        "dist/rpc/redis/index.d.ts"
      ],
      "rabbitmq": [
        "dist/rpc/rabbitmq/index.d.ts"
      ]
    }
  },
  "keywords": [
    "server",
    "express",
    "types-share",
    "fullstack"
  ],
  "author": "",
  "files": [
    "dist",
    "register",
    "bin"
  ],
  "license": "MIT",
  "devDependencies": {
    "@koa/router": "^12.0.1",
    "@nucleoidjs/kafkajs-mock": "^0.0.1",
    "@swc-node/core": "^1.10.6",
    "@types/amqplib": "^0.10.1",
    "@types/debug": "^4.1.12",
    "@types/express": "^4.17.17",
    "@types/koa": "^2.13.12",
    "@types/koa__router": "^12.0.4",
    "@types/supertest": "^2.0.12",
    "amqplib": "^0.10.3",
    "chokidar": "^3.5.3",
    "express": "^4.18.2",
    "fast-glob": "^3.3.2",
    "fastify": "^4.25.1",
    "h3": "^1.9.0",
    "ioredis": "^5.3.2",
    "ioredis-mock": "^8.9.0",
    "kafkajs": "^2.2.4",
    "koa": "^2.14.2",
    "koa-body": "^6.0.1",
    "mock-amqplib": "^1.8.2",
    "supertest": "^6.3.3",
    "tsup": "^6.5.0",
    "typescript": "^5.2.2",
    "unimport": "^3.7.1",
    "vite": "^4.0.0"
  },
  "dependencies": {
    "debug": "^4.3.4",
    "phecda-core": "workspace:*",
    "picocolors": "^1.0.0",
    "reflect-metadata": "^0.1.13"
  }
}
