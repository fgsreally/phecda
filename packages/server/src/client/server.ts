import type amqplib from 'amqplib'
export * from './axios'

export function createMqReq(channel: amqplib.Channel): <R>(arg: R) => Promise<void> {
  return async (arg: any) => {
    const { url, body } = arg
    await channel.assertQueue(url)
    await channel.sendToQueue(url, Buffer.from(JSON.stringify(body)))
  }
}
