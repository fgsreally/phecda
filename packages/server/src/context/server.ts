import type { ValidatePipe } from '../pipe'
import { defaultPipe } from '../pipe'
import type { ErrorFilter } from '../filter'
import { serverFilter } from '../filter'
import { WrongMetaException } from '../exception/wrong-meta'
import { Pcontext } from './base'

export class ServerContext extends Pcontext {
  static pipe = defaultPipe
  static filter = serverFilter
  static middlewareRecord: Record<string, (...params: any) => any> = {}
  static useMiddleware(middlewares: string[]) {
    return middlewares.map((m) => {
      if (!(m in ServerContext.middlewareRecord))
        throw new WrongMetaException(`can't find middleware named ${m}`)
      return ServerContext.middlewareRecord[m]
    })
  }

  async usePipe(args: { arg: any; validate?: boolean }[], reflect: any[]) {
    return ServerContext.pipe.transform?.(args, reflect)
  }

  useFilter(arg: any) {
    return ServerContext.filter(arg, this.data)
  }
}

export function addMiddleware(key: string, handler: (...params: any) => any) {
  ServerContext.middlewareRecord[key] = handler
}

export function useServerPipe(pipe: ValidatePipe) {
  ServerContext.pipe = pipe
}
export function useServerFilter(filter: ErrorFilter) {
  ServerContext.filter = filter
}
