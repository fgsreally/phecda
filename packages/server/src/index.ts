export * from './context'
export * from './types'
export * from './core'
export * from './decorators'
export * from './exception'
export * from './pipe'
export * from './meta'
export * from 'phecda-core'
export * from './server/helper'
export * from './common'
export * from './modules'
export * from './compiler'
export * from './utils'
