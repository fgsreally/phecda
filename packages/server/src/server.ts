import { HttpException } from './exception/base'
import { UndefinedException } from './exception/undefine'
import type { Meta } from './meta'
import type { PhecdaPipe } from './pipe'
import { defaultPipe } from './pipe'

export class PhecdaServer {
  method: string
  params: string[]
  static pipe: PhecdaPipe = defaultPipe
  static guardsRecord: Record<string, (...params: any) => boolean> = {}
  static interceptorsRecord: Record<string, (...params: any) => any | ((...params: any) => any)> = {}
  static serverRecord: Record<string, PhecdaServer> = {}

  constructor(public key: string, public meta: Meta) {
    PhecdaServer.serverRecord[key] = this
  }

  static registerGuard(key: string, handler: any) {
    PhecdaServer.guardsRecord[key] = handler
  }

  static registerInterceptor(key: string, handler: any) {
    PhecdaServer.interceptorsRecord[key] = handler
  }

  async useGuard(req: any, guards: string[]) {
    for (const guard of guards) {
      if (!await PhecdaServer.guardsRecord[guard]?.(req))
        throw new Error('aa')
    }
  }

  async useInterceptor(req: any, interceptors: string[]) {
    const ret = []
    for (const interceptor of interceptors) {
      const post = await PhecdaServer.interceptorsRecord[interceptor]?.(req)
      if (post)
        ret.push(post)
    }
    return ret
  }

  async usePost(ret: any, cbs: ((...params: any[]) => any)[]) {
    for (const cb of cbs)
      ret = (await cb(ret)) | ret

    return ret
  }

  async usePipe(args: { arg: any; validate: boolean }[], reflect: any[]) {
    return PhecdaServer.pipe.transform(args, reflect)
  }

  methodToHandler(method: (...params: any[]) => any) {
    const { data: { params = [], guards = [], interceptors = [] }, reflect = [] } = this.meta

    return async (req: any) => {
      try {
        await this.useGuard(req, guards)
        const posts = await this.useInterceptor(req, interceptors!)
        const args = params.map((param) => {
          return { arg: req[param.type]?.[param.key], validate: param.validate }
        })

        const ret = await method(...await this.usePipe(args, reflect))
        return this.usePost(ret, posts)
      }
      catch (e: any) {
        if (!(e instanceof HttpException))
          return new UndefinedException(e.message || e)
        return e
      }
    }
  }
}

export function addGuard(key: string, handler: (...params: any) => boolean) {
  PhecdaServer.registerGuard(key, handler)
}

export function addInterceptor(key: string, handler: (...params: any) => any | ((...params: any) => any)) {
  PhecdaServer.registerInterceptor(key, handler)
}

export function usePipe(pipe: PhecdaPipe) {
  PhecdaServer.pipe = pipe
}
