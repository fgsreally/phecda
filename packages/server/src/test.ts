import type { Server } from 'node:http'
import type { SuperAgentTest, Test } from 'supertest'
import { type Construct, getTag } from 'phecda-core'
import { Factory } from './core'

import type { PickFunc } from './types'
import type { ControllerMeta } from './meta'
export async function TestFactory<T extends Construct[]>(...Modules: T) {
  const { moduleMap, modelMap } = await Factory(Modules)

  return {
    get<C extends T[number]>(Model: C): InstanceType<C> {
      const tag = getTag(Model)
      const module = moduleMap.get(tag)

      if (!module)
        throw new Error(`module "${String(tag)}" doesn't exist`)

      if (modelMap.get(module) !== Model)
        throw new Error(`module "${Model.name}" and "${String(tag)}" in modulemap are different modules`)

      return module
    },
  }
}

export type SuperTestRequest<T> = {
  [K in keyof T]: T[K] extends (...args: infer R) => any ? (...args: R) => Test : never;
}

export async function TestHttp(app: Server | any, { moduleMap, meta }: Awaited<ReturnType<typeof Factory>>, isAgent = true) {
  const { default: request, agent } = await import('supertest')
  const Agent = agent(app) as SuperAgentTest & { module: typeof module }

  function module<T extends Construct>(Module: T): SuperTestRequest<PickFunc<InstanceType<T>>> {
    const tag = getTag(Module)

    const module = moduleMap.get(tag)
    if (!module)
      throw new Error(`module "${String(tag)}" doesn't exist`)

    return new Proxy({}, {
      get(_target, p) {
        const { data } = (meta as ControllerMeta[]).find(({ data }) => data.name === Module.name && data.func === p && data.tag === tag)!
        return (...args: any) => {
          const ret = { body: {}, headers: {}, query: {}, func: data.http!.type, url: data.http!.prefix + data.http!.route } as any

          data.params.forEach((item) => {
            if (item.type === 'params') {
              ret.url = ret.url.replace(`:${item.key}`, args[item.index])
              return
            }
            if (item.type === 'query') {
              ret.query[item.key] = args[item.index]
              return
            }

            // body
            if (item.key)
              ret[item.type][item.key] = args[item.index]
            else
              ret[item.type] = args[item.index]
          })
          // @ts-expect-error @todo miss type
          let agent = (isAgent ? Agent : request(app))[ret.func](ret.url)
          if (Object.keys(ret.query).length > 0)
            agent = agent.query(ret.query)

          if (Object.keys(ret.headers).length > 0)
            agent = agent.set(ret.headers)

          if (Object.keys(ret.body).length > 0)
            agent = agent.send(ret.body)

          return agent
        }
      },
    }) as any
  }

  Agent.module = module

  return Agent
}
