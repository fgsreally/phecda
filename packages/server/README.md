# phecda-server
types share between server and client, `nestjs` format

## express

actually, it works as a set of middlewares
### server side
```ts
// in test.controller.ts
import { Body, Controller, Get, Param, Post, Query, Watcher, emitter } from 'phecda-server'

@Controller('/base')
export class TestController {
  @Post('/:test')
  async test(@Param('test') test: string, @Body('name') name: string, @Query('id') id: string) {
    console.log(`${test}-${name}-${id}`)
    emitter.emit('watch', 1)
    return `${test}-${name}-${id}`
  }

  @Get('/get')
  async get() {
    return {
      data: 'test',
    }
  }
}
```
```ts
// in server.ts
import fs from 'fs'
import { Factory, bindApp } from 'phecda-server'
import express from 'express'
import { TestController } from './test.controller'
const data = await Factory([TestController])
fs.writeFileSync('meta.p.js', JSON.stringify(data.meta.map(item => item.data)))
const app = express()
app.all('*', (req, res, next) => {
  res.header('Access-Control-Allow-Origin', '*')
  res.header('Access-Control-Allow-Headers', 'Content-Type')
  res.header('Access-Control-Allow-Methods', '*')
  next()
})
app.use(express.json())

bindApp(app, data)
```

### client side

```ts
// in vite.config.ts

import { defineConfig } from 'vite'
import { Server } from 'phecda-server'

export default defineConfig({
  plugins: [Server('meta.p.js')],

})
```

```ts
// in  main.ts

import { $S, createBeacon, createMergeReq, createReq, isError } from 'phecda-server/client'
import axios from 'axios'
import { TestController } from './test.controller'

const instance = axios.create({
  // ...
})
const useRequest = createReq(instance)
const useMergeRequest = createMergeReq(instance)
const { test, get } = new TestController()
async function request() {
  const { data } = await useRequest(test('phecda', 'server', '1'))
  console.log('[normal request]:')

  console.log(data)
}

async function seriesRequest() {
  const { data: [, res2] } = await useMergeRequest([get(), test($S(0, 'data'), 'server', '1')])
  console.log('[series request]:')

  if (isError(res2))
    console.error(res2.message)
  else console.log(res2)
}

async function mergeRequest() {
  const { data: [res1, res2] } = await useMergeRequest([test('phecda', 'server', '1'), get()])
  console.log('[merge request]:')

  if (isError(res1))
    console.error(res1.message)
  else console.log(res1)

  if (isError(res2))
    console.error(res2.message)
  else console.log(res2)
}

request()
mergeRequest()
seriesRequest()
```