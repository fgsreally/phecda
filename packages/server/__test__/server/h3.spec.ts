import { describe, expect, it, vi } from 'vitest'
import request from 'supertest'
import { createApp as createH3, createRouter, toNodeListener } from 'h3'
import type { H3Ctx } from '../../src/server/h3'
import type { HttpOptions } from '../../src'

import { bind } from '../../src/server/h3'
import { ERROR_SYMBOL, Factory, addFilter, addGuard, addInterceptor, addPipe, addPlugin } from '../../src'
import { Test } from '../fixtures/test.controller'

async function createServer(opts?: HttpOptions) {
  const data = await Factory([Test])
  const app = createH3()
  const router = createRouter()
  bind(router, data, opts)
  app.use(router)
  return toNodeListener(app)
}

describe('h3 ', () => {
  it('basic request', async () => {
    const app = await createServer()
    const res1 = await request(app).get('/get')
    expect(res1.body.msg).toBe('test')
    const res2 = await request(app).post('/post/phecda?id=1').send({ name: 'server' })
    expect(res2.text).toBe('phecda-server-1')

    const res3 = await request(app).post('/__PHECDA_SERVER__').send([
      {
        tag: 'Test',
        func: 'post',
        args: ['phecda', 'server', '1'],
      }, {
        tag: 'Test',
        func: 'get',

        args: [],
      },

    ])
    expect(res3.body[0]).toEqual('phecda-server-1')
    expect(res3.body[1]).toEqual({ msg: 'test' })
  })

  it('filter', async () => {
    const app = await createServer()

    addFilter('test', (e) => {
      return { ...e.data, filter: true }
    })
    // default filter
    const res1 = await request(app).get('/error?msg=test')
    expect(res1.body).toEqual({ description: 'Exception', message: 'test', status: 500, [ERROR_SYMBOL]: true })
    // specfic filter--test
    const res2 = await request(app).get('/filter')
    expect(res2.body).toEqual({ description: 'Exception', message: 'filter error', status: 500, [ERROR_SYMBOL]: true, filter: true })
  })
  it('pipe', async () => {
    const app = await createServer()

    addPipe('add', ({ arg }) => {
      return arg + 1
    })

    // default pipe
    const res1 = await request(app).post('/pipe').send({ info: { name: '' } })

    expect(res1.body).toMatchObject({ message: 'name should be phecda', [ERROR_SYMBOL]: true })
    // specfic pipe--add
    const res2 = await request(app).post('/pipe').query({ id: '1' }).send({ info: { name: 'phecda' } })
    expect(res2.text).toBe('11-phecda')
  })

  it('global filter/pipe', async () => {
    const fn = vi.fn()
    addPipe('express', ({ arg }) => {
      fn()
      return arg
    })
    addFilter('express', (e) => {
      fn()
      return e.data
    })
    const app = await createServer({ globalPipe: 'express', globalFilter: 'express' })
    await request(app).get('/error?msg=test')
    expect(fn).toBeCalledTimes(2)
  })

  it('plugin', async () => {
    const fn = vi.fn()

    const plugin = () => {
      return () => {
        fn()
      }
    }
    addPlugin('p1', plugin)
    addPlugin('p2', plugin)

    const app = await createServer({ parallelPlugins: ['p1'], globalPlugins: ['p2'] })

    await request(app).get('/plugin')
    expect(fn).toHaveBeenCalledTimes(2)
    await request(app).post('/__PHECDA_SERVER__').send([{
      tag: 'Test',
      func: 'plugin',
      args: [],
    }])

    expect(fn).toHaveBeenCalledTimes(4)
  })

  it('guard/interceptor', async () => {
    const InterceptFn = vi.fn((str: string) => str)
    const Guardfn = vi.fn((str: string) => str)

    addGuard('g1', ({ params, index }: H3Ctx) => {
      Guardfn('g1')
      if (index === undefined && params.test !== 'test')
        return false
      return true
    })
    addGuard('g2', () => {
      Guardfn('g2')
      return true
    })

    const mockInterceptor = (ctx: H3Ctx) => {
      InterceptFn('start')
      return () => {
        InterceptFn('end')
        if (!ctx.parallel)
          return { ps: true }
      }
    }
    addInterceptor('i1', mockInterceptor)
    addInterceptor('i2', mockInterceptor)

    const app = await createServer({
      globalGuards: ['g1'],
      globalInterceptors: ['i1'],
    })
    const res1 = await request(app).post('/aop/no')
    expect(res1.body).toMatchObject({
      message: 'Guard exception--[g1]',
      status: 403,
      [ERROR_SYMBOL]: true,
    })
    expect(Guardfn).toHaveBeenCalledTimes(1)
    expect(InterceptFn).toHaveBeenCalledTimes(0)

    await request(app).post('/aop/test').expect(200, { ps: true })

    expect(Guardfn).toHaveBeenCalledTimes(3)
    expect(InterceptFn).toHaveBeenCalledTimes(3)

    await request(app).post('/__PHECDA_SERVER__').send(
      [
        {
          tag: 'Test',
          func: 'aop',
          args: ['test1'],
        },
        {
          tag: 'Test',
          func: 'aop',

          args: ['test2'],
        },
      ],

    )

    expect(InterceptFn).toHaveBeenCalledTimes(11)

    expect(Guardfn).toHaveBeenCalledTimes(7)
  })
  it('ctx', async () => {
    addGuard('g', (ctx: H3Ctx) => {
      expect({ body: ctx.body, query: ctx.query, params: ctx.params }).toMatchSnapshot()
      return true
    })

    const app = await createServer({
      globalGuards: ['g'],
    })

    await request(app).post('/all/test?id=1').send({ name: 'test' }).expect(200, ['test', { name: 'test' }, '1'])
    await request(app).post('/__PHECDA_SERVER__').send(
      [
        {
          tag: 'Test',
          func: 'all',

          args: ['test', { name: 'test' }, '1'],
        },
        {
          tag: 'Test',
          func: 'all',

          args: ['test', { name: 'test' }, '2'],
        },
      ],

    ).expect(200)
  })
})
