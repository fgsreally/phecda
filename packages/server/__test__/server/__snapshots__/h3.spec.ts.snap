// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`h3  > ctx 1`] = `
{
  "body": {
    "name": "test",
  },
  "params": {
    "test": "test",
  },
  "query": {
    "id": "1",
  },
}
`;

exports[`h3  > ctx 2`] = `
{
  "body": {
    "name": "test",
  },
  "params": {
    "test": "test",
  },
  "query": {
    "id": "1",
  },
}
`;

exports[`h3  > ctx 3`] = `
{
  "body": {
    "name": "test",
  },
  "params": {
    "test": "test",
  },
  "query": {
    "id": "2",
  },
}
`;
